Fix reference updated concurrently error for the filesystem storer

Javier-varez opened the pull request:
If a reference exists in the packed refs file but not in the actual refs folder, we need to fall back to reading the pack refs file when storing the reference.

This happens because by the time we are storing the reference, we open the file with write permissions and read from it. At that point, if the file is empty, we still need to read the valid reference. Otherwise, it will read 0, and assume that the references do not match (because the old value was read from the packed file).

Relates to https://github.com/go-git/go-git/issues/37

pjbgf commented on storage/filesystem/dotgit/dotgit.go:
```diff
@@ -655,24 +655,47 @@ func (d *DotGit) ObjectDelete(h plumbing.Hash) error {
 	}
+
 	ref, err := d.readReferenceFrom(f, old.Name().String())
+	if err == emptyRefFileError {
```
```suggestion
	if errors.Is(err, ErrEmptyRefFile) {
```

pjbgf commented on storage/filesystem/dotgit/dotgit.go:
```diff
@@ -655,24 +655,47 @@ func (d *DotGit) ObjectDelete(h plumbing.Hash) error {
 	return err1
 }
 
+type refError string
+
+func (r refError) Error() string {
+	return string(r)
+}
+
+var emptyRefFileError = refError("The file is empty")
```
Please move this to the top together with the other errors.

```suggestion
var ErrEmptyRefFile = errors.New("ref file is empty")
```

pjbgf commented on storage/filesystem/dotgit/dotgit.go:
```diff
@@ -655,24 +655,47 @@ func (d *DotGit) ObjectDelete(h plumbing.Hash) error {
 	}
 
+	if len(b) == 0 {
+		return nil, emptyRefFileError
```
```suggestion
		return nil, ErrEmptyRefFile
```

pjbgf reviewed (changes requested):
@Javier-varez nice stuff! Thanks for looking into this issue. Apart from the nits below this LGTM.

Javier-varez commented:
> @Javier-varez nice stuff! Thanks for looking into this issue. Apart from the nits below this LGTM.

Thank you for reviewing the PR. I just addressed the comments 👍🏻

pjbgf reviewed (approved):
@Javier-varez thanks for working on this. :bow:
